src/dhtproto/node/neo/request/Update.d

```diff
@@ -0,0 +1,329 @@
+
+        MessageType response;
+
+        if ( this.responsibleForKey(key) )
+        {
+            with ( MessageType ) switch ( message )
+            {
+                // Normal sequence: read, wait for response, update.
+                case GetRecord:
+                    response = this.getWaitUpdate(channel, key);
+                    break;
+
+                // Either an updated record transferred from another node or a
+                // new record that did not exist in the DHT before. Just put it.
+                case UpdateRecord:
+                    response = this.update(channel, key, this.init_payload);
+                    break;
+
+                default:
+                    ed.shutdownWithProtocolError("Unexpected message from client");
+            }
+        }
+        else
+            response = MessageType.WrongNode;
+
+        // Send the final message to the client.
+        ed.send(
+            ( ed.Payload payload )
+            {
+                payload.add(response);
+            }
+        );
+        ed.flush();
+    }
+
+    /***************************************************************************
+
+        Normal sequence: get, wait for response, update.
+
+        Params:
+            channel = channel to update record in
+            key = key of record to update
+
+        Returns:
+            message type code to return to the client
+
+    ***************************************************************************/
+
+    private MessageType getWaitUpdate ( cstring channel, hash_t key )
+    {
+        auto ed = this.connection.event_dispatcher();
+
+        // Get record value from storage and send it to the client.
+        bool exists;
+        auto success = this.get(channel, key,
+            ( Const!(void)[] value )
+            {
+                exists = true;
+
+                ed.send(
+                    ( ed.Payload payload )
+                    {
+                        payload.addCopy(MessageType.RecordValue);
+                    }
+                );
+                ed.flush();
+
+                // TODO: The value could be sent in the same message as the
+                // response. For simplicity, we send them separately, at
+                // present.
+                ed.send(
+                    ( ed.Payload payload )
+                    {
+                        payload.addArray(value);
+                    }
+                );
+                ed.flush();
+            }
+        );
+
+        // If getting the record failed, end the request.
+        if ( !success )
+            return MessageType.Error;
+
+        // If the record did not exist, end the request.
+        if ( !exists )
+        {
+            ed.send(
+                ( ed.Payload payload )
+                {
+                    payload.addCopy(MessageType.NoRecord);
+                }
+            );
+            ed.flush();
```

Comment:
Why not just `return MessageType.NoRecord`?

Reply:
It looks like you don't want to end the request, the code comment is wrong then.

Reply:
It could be left to the client to decide what to do in this case and respond either with a record value to put in the DHT or a message to cancel the request. This would increase the complexity of the request but be user friendly, avoiding a subsequent `Put` request if the record doesn't exist.

Reply:
Yeah, I deliberately chose to leave this as a future enhancement. (An earlier version of Update actually did handle this extra logic, but I removed it to simplify things. I guess this is where this bug / inconsistency comes from.)